### DISCUSSION TITLE
suggesting a better naming rules for images saved in freeplane map file (using timestamp in the filenames instead of random naming)

### DISCUSSION THREAD

**Discussion by Bricklam** [freeplane/freeplane#1086] [Ideas]
when I copy opened image file into freeplane, freeplane will create a copy of the image file with a different name and place it in a folder associated with the map file.
(a folder that have the same filename as the map file + the word "_files" )

the problem is: freeplane would give every pasted/generated image a random filename, that do not identify the image file creation date/time.
So, the user wouldn't know the order of the images in the map associated folder (which image is pasted/generated before the other ones)

here are some examples of the default filenames given by freeplane:
"D:\New Mindmap_files\png_959866522589316445.png"
"D:\New Mindmap_files\png_975582140529345501.png"

my suggesion is: to give every image a filename with a timestamp (date/time) by default, to mark their creation date/time,
 to maintain their order in the associated folder based on creation date/time (pasting time),
 to be in the same order of adding them to freeplane.
So, instead of using random filenames, the timestamped filename can have these variables:
 (Year-Month-Day, Hour-Minute-Second-Millisecond)

for example:
"D:\New Mindmap_files\png_2023-04-09_02-44-13.png"
"D:\New Mindmap_files\png_2023-04-09_02-45-77.png"

This naming would have a big advantage: when the user try to view the images in the mind map associated folder (_files),
using external image viewer app, he/she can browse the images in the same order as they were pasted to Freeplane

Best regards

**Comment by EdoFro** [comment]
> So, instead of using random filenames, the timestamped filename can have these variables:
(Year-Month-Day, Hour-Minute-Second-Millisecond)
> 
> for example: 
"D:\New Mindmap_files\png_2023-04-09_02-44-13.png" 
"D:\New Mindmap_files\png_2023-04-09_02-45-77.png"

I like your proposal, but I would modify it a little:
- img-YearMonthDay-HourMinute-NodeID.ext

for example:
"D:\New Mindmap_files\img-20230409-1445-ID_52900898.png"

because that way, if you are looking at the images in your folder and want to see where it was pasted in the map, you can go to the map and search the node by its ID.

jm2c
edo

**Reply by Illegit** [reply]
Is this a setting we have to do or a sw configuration? How to make it run this "img-YearMonthDay-HourMinute-NodeID.ext"?
Thanks

**Reply by Bricklam** [reply]
nice idea, I agree

**Comment by beccon4** [comment]
In the fear of repeating myself: why not embed the image files into the mm file?

It would make things so much simpler and more convenient. Things like sharing a Mindmap with soneone else or to use Freeplane as a shared documentation and/or presentation tool. 

I've failed multiple times to introduce Freeplane due to the fact that images etc. get saved outside the main file and have to be considered when sending the mindmap. Platform diversity - here Linux, there Windows, next one on MacOSX and the increasing use document management systems (Sharepoint) etc. does not make things easier. 

Probably, instead of trying to improve linking it is time to take a step back and take a different approach. :-)

There used to be an addon "Pack and go" but this a) has to be installed b) does not work anymore with recent Freeplane versions.

**Reply by Bricklam** [reply]
I don't think it is a good idea to embed images directly into Freeplane
I'm a fraid that would slow it down and reduce its performance, specially you have many images embedded, which is too bad since speed & performance should always have the top priority.

linking to external images can work (even between different systems), if you have everything set correctly (using relative paths),
and you can archive the entire freeplane map folder as (zip) file (the map + the associated folder) before sending it

for the add-on "Pack and go", I think it was a great tool. it is unfortunate that it doesn't work anymore,
but I think it can be updated or even re-designed to work with recent Freeplane versions.
if we can call the developer of "Pack and go" to check what is wrong with it
 I think that might solve the problem

there is a note taking application that allow embedding images inside it: (Cherrytree),
 but it is very different. it has a binary database format, unlike Freeplane XML format, 

thanks

**Reply by beccon4** [reply]
Performance could be an issue - especially since we're on the Java platform. So we should definitely keep the option to link files - just because not everything we have is images - there is PDF and even office documents. 

But I'd stick to my use case: take Freeplane for documentation, knowledge sharing and non boring, non Powerpoint presentations which all require sending the file to someone else who is most likely non techie. As I'm a seasoned Linuxer and I know about the differences between files system and their structure. Please look above and see the fully qualified Windows file names. They would not work on a Mac.

An example: In a recent telco with the folks of my Neighbourhood Improvement Scheme I'm working with, I presented a mindmap with all the hotspots around here and we collected ideas on what to do about it. Later I wanted to send off the map to the participants who are teachers, shop owners, pensioners - what ever - the usual suspects.  They might install Freeplane on ther shiny Macbooks and use Freeplane even for their other activities - but I can't tell them where to find the images to put things together again. They would just reject the concept for beeing too complicated. They would expect a file in an email, click on it and that's it.

Next Example: I held an online presentation about crypto currencies. Audience: bank folks, freedom activists and the usual "digital as a foreign language" crowd. For that I learned the really nice presentation function of Freeplane - but as to keep the audience awake I should illustrate my points - which is images. The same thing: send the map to the participants ("I have a Mac", "I do not know how th get to view the images" - same procedure.

Third example: Freemind as a documentation and knowledge sharing tool. Since I'm freelancing - i.e. work as a consultant who enters a project at some stage, collects information and then help to find solutions - I use Freeplane (actually even before it came out - Freemind was the choice) Increasingly I share this information with colleagues - so we send it back and forth (for the lack of an online Freeplane version) The issue here: what to do with the images included. New ones get included - some changed -some disappear. Images here include screen shots and photos taken from the white board (the physical one). The same issue.

I'd doubt that a couple of jpegs in a map would crash a modern system with some GB of RAM. And as scaling libraries are in the code anyway (for display) - one can use this as an option too.

Having the possibility to embed would make Freeplane so much more versatile and usable for non technical people ... Freeplane is brilliant and very useful for creative and visually minded people - but to excel it should be easier to handle.

**Reply by beccon4** [reply]
Maybe one could take a PackToGo approach: as we already use a temporary folder to store the snapshots in, it could be used to write out image files. But unlike zipping everything (which does not save space with already compressed jpegs) use the component structure of the XML file.

**Comment by dpolivaev** [comment]
I think if you want to contribute and if you are interested in this feature you could do it yourself. Let me know if you need any assistance.

**Reply by Bricklam** [reply]
Yes. this is a great idea @dpolivaev  and it is an easy one to start from.  since I didn't know exactly where to start.
OK. I will check the relevant code files and return back with my findings and my suggestions (as soon as I come up with reasonable solutions to implement this feature .)
Thanks
